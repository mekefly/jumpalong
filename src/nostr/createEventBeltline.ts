import { EventBeltline, EventBeltlineOptions } from "./eventBeltline";

export function createEventBeltline(
  options?: EventBeltlineOptions & {
    addExtendsFromParent?: boolean;
    addExtendsFormRoot?: boolean;
  }
): EventBeltline<{}> {
  return EventBeltline.root.createChild(Object.assign({}, options));
}
export function createEventBeltlineReactive(
  options?: EventBeltlineOptions & {
    addExtendsFromParent?: boolean;
    addExtendsFormRoot?: boolean;
  }
): EventBeltline<{}> {
  return EventBeltline.root.createChild(
    Object.assign({ slef: reactive({}) }, options)
  );
}
