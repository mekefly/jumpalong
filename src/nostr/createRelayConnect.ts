import { Relay } from "nostr-tools";

export function createRelayConnect(relay: Relay) {
  return { relay, taskCount: 0 };
}
