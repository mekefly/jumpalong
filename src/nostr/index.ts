import { actors } from "../actor/actorSystem";
import { RelayQuery } from "./relay";
export const relayQuery = actors().then((rootActor) => {
  return rootActor.createChild(RelayQuery, { clusterSize: 20 });
});
