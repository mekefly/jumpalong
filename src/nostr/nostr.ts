import { EventBeltline } from "./eventBeltline";
import { RelayEmiter } from "./RelayEmiter";
import { RelayPool } from "./RelayPool";

export function initializeCore() {
  const relayEmiter = new RelayEmiter();
  const relayPool = new RelayPool();
  const eventBeltline = new EventBeltline({
    preventCircularReferences: true,
    relayEmiter,
  });
  // .addStaff(createLocalStorageStaff())
  // .addStaff(
  //   UseStorageStaff(
  //     [
  //       {
  //         kinds: [1, 42],
  //       },
  //     ],
  //     300
  //   )
  // )
  // .addStaff(
  //   UseStorageStaff(
  //     [
  //       {
  //         kinds: [0, 40, 41],
  //       },
  //     ],
  //     500
  //   )
  // );

  relayEmiter.onEvent(({ subId, event }) => {
    eventBeltline.pushEvent(event, subId);
  });

  return { relayEmiter, relayPool, eventBeltline };
}
export const { relayEmiter, relayPool, eventBeltline } = initializeCore();
