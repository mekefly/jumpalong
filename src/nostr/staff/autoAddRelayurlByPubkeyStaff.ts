import { createStaff, StaffThisType } from ".";
import root from "../eventBeltline";
import createAutomaticRandomRequestStaff from "./automaticRandomRequestStaff";
import createAddRelayUrlGraspClues from "./createAddRelayUrlGraspClues";
import createEoseUnSubStaff from "./createEoseUnSubStaff";
import { createLatestEventStaff } from "./createLatestEventStaff";
import createReadWriteListStaff from "./createReadWriteListStaff";

export type AddRelayurlByPubkeyStaff = {
  initialization(this: StaffThisType<{}>): void;
};

/**
 * 自动添加用户配置的url到生产线，通过用户公钥
 * @param pubkey
 * @returns
 */
export default function autoAddRelayurlByPubkeyStaff(
  pubkey: string,
  opt?: {}
): AddRelayurlByPubkeyStaff {
  return createStaff({
    initialization() {
      console.log("autoAddRelayurlByPubkeyStaff:initialization");

      const slefBeltline = this.beltline;
      const kind10002line = slefBeltline
        .createChild({
          addExtendsFromParent: false,
          addExtendsFormRoot: true,
        })
        //10002是用户建议的relay读写列表
        .addFilter({
          kinds: [10002],
          authors: [pubkey],
        })
        .addStaff(createLatestEventStaff())
        .addStaff(createEoseUnSubStaff())
        .addStaff(createReadWriteListStaff())
        .addStaff(createAddRelayUrlGraspClues())
        .addExtends(root);

      kind10002line.feat.onHasReadWriteList((readWrite) => {
        slefBeltline.addRelayUrls(readWrite.writeUrl);
      });

      setTimeout(() => {
        //缓存返回true，认为是有效的缓存，在一段时间内只允许请求一次

        //请求到至少一次metadata
        if (kind10002line.feat.isHas()) {
          return;
        }
        kind10002line.addReadUrl();
        //如果两秒后还没获取到需要的数据，就会自动扫描中继
        const kind10002line1 = kind10002line.addStaff(
          createAutomaticRandomRequestStaff()
        );
        //请求到至少一次metadata
        kind10002line1.feat.onHasLatestEventOnce(() => {
          kind10002line1.feat.stopAutomaticRandomRequestStaff();
        });
      }, 100);
    },
  });
}
