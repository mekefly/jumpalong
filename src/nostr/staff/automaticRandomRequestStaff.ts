import { relayConfigurator } from "@/api/relayConfigurator";
import { Filter } from "nostr-tools";
import { createStaff, StaffThisType } from ".";
import root from "../eventBeltline";
import createEoseUnSubStaff from "./createEoseUnSubStaff";

export type AutomaticRandomRequestStaff = {
  initialization(this: StaffThisType<{}>): void;
  feat: AutomaticRandomRequestStaffFeat;
};
export type AutomaticRandomRequestStaffFeat = {
  startAutomaticRandomRequestStaff: () => void;
  stopAutomaticRandomRequestStaff: () => void;
};
class AutoRandomRequestStaff {
  interval = 3_000;
  maximumTimes = 30;
  setToBeAdded = new Set<string>();
  added = new Set();
  count = 0;
  timer: NodeJS.Timer | undefined = undefined;

  filterMap = new Map<string, Filter>();

  stop() {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  startProcessing() {
    this.count = 0;

    //已经运行阻止再次运行
    if (this.timer) return;
    this.timer = setInterval(() => {
      this.count++;

      let size = this.setToBeAdded.size;

      if (size <= 0) {
        this.setToBeAdded = new Set(relayConfigurator.getOtherList());
        size = this.setToBeAdded.size;
      }

      //工人是懒惰的，干一定量的活就偷偷打盹
      if (
        size <= 0 ||
        this.filterMap.size <= 0 ||
        this.count > this.maximumTimes
      ) {
        this.stop();
        return;
      }

      //随机index
      const index = Math.floor(Math.random() * size);
      const randomUrl = Array.from(this.setToBeAdded)[index];

      // 维护已添加列表和等待添加的列表
      this.added.add(randomUrl);
      this.setToBeAdded.delete(randomUrl);

      const filters = Array.from(this.filterMap, ([k, v]) => v);

      // 请求数据
      root
        .addFilters(filters)
        .addStaff(createEoseUnSubStaff())
        .addRelayUrls(new Set<string>().add(randomUrl));
    }, this.interval);
  }

  addFilter(filters: Filter[]) {
    for (const filter of filters) {
      autoRandomRequestStaff.filterMap.set(JSON.stringify(filter), filter);
    }
    autoRandomRequestStaff.startProcessing();
  }
  removeFilters(filters: Filter[]) {
    if (!filters) return;

    for (const filter of filters) {
      autoRandomRequestStaff.filterMap.delete(JSON.stringify(filter));
    }
  }
}
export const autoRandomRequestStaff = new AutoRandomRequestStaff();

//随缘算法

/**
 * 自动随机添加url，主要用于查找用户所在的中继线索的时候使用
 * @param opt
 * @returns
 */
export default function createAutomaticRandomRequestStaff() {
  let filters: Filter[] | null = null;

  function stopAutomaticRandomRequestStaff() {
    if (!filters) return;
    autoRandomRequestStaff.removeFilters(filters);
  }

  return createStaff({
    initialization() {
      (this.beltline.feat as any).startAutomaticRandomRequestStaff();
    },
    feat: {
      startAutomaticRandomRequestStaff() {
        const filters = this.beltline.getFilters();
        autoRandomRequestStaff.addFilter(filters);
      },
      stopAutomaticRandomRequestStaff,
    },
    onClose() {
      stopAutomaticRandomRequestStaff();
    },
  });
}
