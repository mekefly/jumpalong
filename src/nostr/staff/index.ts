import { Event } from "nostr-tools";
import { StaffState } from "./Staff";
export * from "./createDoNotRepeatStaff";
export * from "./createFilterGreaterThanTheCurrenttimeStaff";
export { createFilterStaff, type FilterStaff } from "./createFilterStaff";
export * from "./createOneEventStaff";
export { default as createOneEventStaff } from "./createOneEventStaff";
export * from "./createPreventCircularReferencesStaff";
export * from "./Staff";


