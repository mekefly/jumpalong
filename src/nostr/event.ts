import { userKey } from "@/nostr/user";
import * as secp256k1 from "@noble/secp256k1";
import { Event, getEventHash, signEvent } from "nostr-tools";

export function validateEvent(event: Event): boolean {
  if (typeof event.content !== "string") return false;
  if (typeof event.created_at !== "number") return false;
  if (typeof event.pubkey !== "string") return false;
  if (!event.pubkey.match(/^[a-f0-9]{64}$/)) return false;

  if (!Array.isArray(event.tags)) return false;
  for (let i = 0; i < event.tags.length; i++) {
    let tag = event.tags[i];
    if (!Array.isArray(tag)) return false;
    for (let j = 0; j < tag.length; j++) {
      if (typeof tag[j] === "object") return false;
    }
  }

  return true;
}
export function verifySignature(event: Event & { sig: string }): boolean {
  return secp256k1.schnorr.verifySync(
    event.sig,
    getEventHash(event),
    event.pubkey
  );
}

export function createEvent(options: Partial<Event>) {
  const { privateKey, publicKey } = userKey.value;
  let event: Event = Object.assign(
    {
      kind: 1,
      pubkey: publicKey,
      created_at: Math.floor(Date.now() / 1000),
      tags: [],
      content: "",
    },
    options
  );

  event.id = getEventHash(event);
  event.sig = signEvent(event, privateKey);
  return event;
}
