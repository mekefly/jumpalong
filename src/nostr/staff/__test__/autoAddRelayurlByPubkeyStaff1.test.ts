import { createEventBeltline } from "@/nostr/createEventBeltline";
import { timeout } from "@/utils/utils";
import { blockRequest } from "@/utils/vitest";
import autoAddRelayurlByPubkeyStaff from "../autoAddRelayurlByPubkeyStaff";
import { autoRandomRequestStaff } from "../automaticRandomRequestStaff";

it("autoAddRelayurlByPubkeyStaff:1", async () => {
  console.log("autoAddRelayurlByPubkeyStaff:1");

  const arr: any[] = [];
  const urls: any[] = [];
  blockRequest(
    {
      req(filter) {
        arr.push(filter);
        return "testId";
      },
    },
    undefined,
    (url) => {
      console.log(url);
      urls.push(url);
    }
  );

  autoRandomRequestStaff.interval = 1;
  autoRandomRequestStaff.maximumTimes = 3;
  autoRandomRequestStaff.setToBeAdded = new Set(
    Array.from({ length: 100 }, (v, index) => `wss://test2.com`)
  );
  console.log(autoRandomRequestStaff.setToBeAdded);

  const line = createEventBeltline()
    .addFilter({
      kinds: [10002],
      authors: [
        '"38440fe4b3729a614592d712e722edd87f7c3bc40e316901ed5b436f25dbcc02"',
      ],
    })
    .addStaff(
      autoAddRelayurlByPubkeyStaff(
        "38440fe4b3729a614592d712e722edd87f7c3bc40e316901ed5b436f25dbcc02"
      )
    );

  await timeout(500);

  expect(arr).toMatchSnapshot();
  expect(urls).toMatchSnapshot();
});
