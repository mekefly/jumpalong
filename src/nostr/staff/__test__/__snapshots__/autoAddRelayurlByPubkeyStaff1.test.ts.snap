// Vitest Snapshot v1

exports[`autoAddRelayurlByPubkeyStaff:1 1`] = `
[
  [
    {
      "authors": [
        "38440fe4b3729a614592d712e722edd87f7c3bc40e316901ed5b436f25dbcc02",
      ],
      "kinds": [
        10002,
      ],
    },
  ],
]
`;

exports[`autoAddRelayurlByPubkeyStaff:1 2`] = `
[
  "wss://test2.com",
]
`;
